Sources/swift-format/Utilities/FileIterator.swift

```diff
@@ -14,54 +14,84 @@ import Foundation
+  private let followSymlinks: Bool
 
   /// Iterator for the list of URLs.
-  var urlIterator: Array<URL>.Iterator
+  private var urlIterator: Array<URL>.Iterator
 
   /// Iterator for recursing through directories.
-  var dirIterator: FileManager.DirectoryEnumerator? = nil
+  private var dirIterator: FileManager.DirectoryEnumerator? = nil
 
   /// The current working directory of the process, which is used to relativize URLs of files found
   /// during iteration.
-  let workingDirectory = URL(fileURLWithPath: ".")
+  private let workingDirectory = URL(fileURLWithPath: ".")
 
   /// Keep track of the current directory we're recursing through.
-  var currentDirectory = URL(fileURLWithPath: "")
+  private var currentDirectory = URL(fileURLWithPath: "")
 
   /// Keep track of files we have visited to prevent duplicates.
-  var visited: Set<String> = []
+  private var visited: Set<String> = []
 
   /// The file extension to check for when recursing through directories.
-  let fileSuffix = ".swift"
+  private let fileSuffix = ".swift"
 
   /// Create a new file iterator over the given list of file URLs.
   ///
   /// The given URLs may be files or directories. If they are directories, the iterator will recurse
   /// into them.
-  init(urls: [URL]) {
+  public init(urls: [URL], followSymlinks: Bool) {
     self.urls = urls
     self.urlIterator = self.urls.makeIterator()
+    self.followSymlinks = followSymlinks
   }
 
   /// Iterate through the "paths" list, and emit the file paths in it. If we encounter a directory,
   /// recurse through it and emit .swift file paths.
-  mutating func next() -> URL? {
+  public mutating func next() -> URL? {
     var output: URL? = nil
     while output == nil {
       // Check if we're recursing through a directory.
       if dirIterator != nil {
         output = nextInDirectory()
       } else {
-        guard let next = urlIterator.next() else { return nil }
-        var isDir: ObjCBool = false
-        if FileManager.default.fileExists(atPath: next.path, isDirectory: &isDir), isDir.boolValue {
-          dirIterator = FileManager.default.enumerator(at: next, includingPropertiesForKeys: nil)
+        guard var next = urlIterator.next() else {
+          // If we've reached the end of all the URLs we wanted to iterate over, exit now.
+          return nil
+        }
+
+        guard
+          let fileType = try? next.resourceValues(forKeys: [.fileResourceTypeKey]).fileResourceType
+        else {
+          continue
+        }
+
+        switch fileType {
+        case .symbolicLink:
+          guard
+            followSymlinks,
+            let destination = try? FileManager.default.destinationOfSymbolicLink(atPath: next.path)
+          else {
+            break
+          }
+          next = URL(fileURLWithPath: destination)
+          fallthrough
+
+        case .directory:
+          dirIterator = FileManager.default.enumerator(
+            at: next,
+            includingPropertiesForKeys: nil,
```

Comment:
You could prefetch the `fileResourceTypeKey` here. I'm not sure how much it'll speed up the iteration, but I think it'll combine the later look-up into whenever the iterator does file IO.

Reply:
Unfortunately it looks like the whole `URL.resourceValues` API is busted on Linux. Switched over to `FileManager.attributesOfItem`, which doesn't have the issue (but also doesn't benefit from pre-fetching).